### PR TITLE
Add ModelEditing pipeline

### PR DESCRIPTION
**Pull request by bahjat-kawar**
This PR adds a StableDiffusionModelEditingPipeline for [Editing Implicit Assumptions in Text-to-Image Diffusion Models](https://time-diffusion.github.io/), a method that allows users to edit the assumptions implicit in Stable Diffusion using an efficient closed-form solution. The edit should affect related prompts' generations while leaving others unaffected.

```python
import torch
from diffusers import StableDiffusionModelEditingPipeline

model_ckpt = "CompVis/stable-diffusion-v1-4"
pipe = StableDiffusionModelEditingPipeline.from_pretrained(model_ckpt)

pipe = pipe.to("cuda")

source_prompt = "A pack of roses"
destination_prompt = "A pack of blue roses"
pipe.edit_model(source_prompt, destination_prompt)

prompt = "A field of roses"
image = pipe(prompt).images[0]
image.save("field_of_roses.png")
```

### CODE REVIEW

**Comment by sayakpaul** [comment]
@bahjat-kawar thanks for your work here. Could you explain (in the interest for the community to understand it better) how it's different from https://huggingface.co/docs/diffusers/main/en/api/pipelines/semantic_stable_diffusion ?

Of course, it's fine for pipelines to do similar things but it's also important for us to be aware of the ways in which they might be different.

**Comment by sayakpaul** [comment]
Thanks for explaining this!

> It allows the editing of the diffusion model weights such that certain assumptions made by the model are changed for all prompts (e.g., turning all roses blue, in all prompt generations).

So, if I understand correctly, after the model weights have been edited to reflect that the roses are blue, all roses will be generated to have blue colors, right?

**Comment by bahjat-kawar** [comment]
> Thanks for explaining this!
> 
> > It allows the editing of the diffusion model weights such that certain assumptions made by the model are changed for all prompts (e.g., turning all roses blue, in all prompt generations).
> 
> So, if I understand correctly, after the model weights have been edited to reflect that the roses are blue, all roses will be generated to have blue colors, right?

Yes, that is the goal :)

File: docs/source/en/_toctree.yml
Comment (sayakpaul) [inline docs/source/en/_toctree.yml:196]: Since it's already under `stable_diffusion`, no need to add it here:

```suggestion
```

File: docs/source/en/api/pipelines/stable_diffusion/model_editing.mdx
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/model_editing.mdx:28]: After the pipeline is merged, would be nice to update the demo with `diffusers`. Maybe @hysts could open a PR :)

File: docs/source/en/index.mdx
Comment (sayakpaul) [inline docs/source/en/index.mdx:92]: Since it's already to Stable Diffusion, it's okay to remove it from here:

```suggestion
```

File: docs/source/en/api/pipelines/stable_diffusion/model_editing.mdx
Comment (sayakpaul) [inline docs/source/en/api/pipelines/stable_diffusion/model_editing.mdx:35]: Do you want to add one / two lines to explain why this kind of model editing is super useful?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:91]: Does it only work with `DDIMScheduler`? Most of the pipelines in diffusers work with different schedulers. Some such as MultiDiffusion don't with stateful schedulers.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:470]: It's not immediately clear what restarting is doing here.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:126]: Shouldn't we test out the editing capability here as well?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:150]: Same comment regarding model editing.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:172]: Same comment regading model editing.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:241]: Excellent!

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:294]: What is the minimum amount required here? 

Can we test if the assertion passes if `mem_bytes < 4.0 * 10**9`?

**Review by sayakpaul — APPROVED** [review]
Excellent work! I just left some nits. 

Appreciate your willingness to contribute this so quickly. Well done 🔥 

I am running the slow tests locally on my end. Will also take care of resolving the merge conflicts.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py
Comment (bahjat-kawar) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:91]: No. This was a mistake. Will fix in next commit, thanks!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:1]: Needs to be changed.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (bahjat-kawar) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:286]: Passed with 4.4 * 10**9. Will update in next commit.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:485]: Should we collect all those hardcoded prompt beginnings in a constants var in the beginning of the script? Also do we want to allow the user to overwrite those or not?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (patrickvonplaten) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:233]: Yes let's delete this test here if possible @bahjat-kawar - our CI is already quite slow, so we should try to not add too many tests. Let's remove this one

File: tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py
Comment (bahjat-kawar) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_model_editing.py:126]: The current model editing code edits layers with the class name "CrossAtention" and input_dim of 768 (like CLIP). These don't exist in the dummy model, so editing would not affect it. The slow tests make sure the editing behavior is valid.

**Comment by bahjat-kawar** [comment]
Thanks @sayakpaul and @patrickvonplaten !
I removed the lms test, added an entry to [diffusers/docs/source/en/index.mdx](https://github.com/huggingface/diffusers/blob/dc5b4e2342432d5efba9692809d49ee13756a2ae/docs/source/en/index.mdx?plain=1#L52), and fixed the issues raised.
Fast tests will not test the editing capabilities because the code is limited to CLIP-based cross-attention. Slow tests cover this, though.
If all checks pass, we should be ready to merge. Thanks for the excellent reviews!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_model_editing.py:32]:
```suggestion
AUGS_CONST = ["A photo of ", "An image of ", "A picture of "]
```

**Review by sayakpaul — APPROVED** [review]
Thanks a lot :) 

I would still like for @patrickvonplaten to give it a quick look regarding the customizability of the prompts used in the `edit_model()` method. 

Maybe expose an option in `edit_model()` -- `aug_prompts`. If it's `None` (which could be made default) we use `AUGS_CONST`. WDYT? 

@bahjat-kawar,

If, in your experience, those prompts suffice and exposing an option to customize them doesn't help with the overall performance, we can discard the possibility.

**Comment by bahjat-kawar** [comment]
> Thanks a lot :)
> 
> I would still like for @patrickvonplaten to give it a quick look regarding the customizability of the prompts used in the `edit_model()` method.
> 
> Maybe expose an option in `edit_model()` -- `aug_prompts`. If it's `None` (which could be made default) we use `AUGS_CONST`. WDYT?
> 
> @bahjat-kawar,
> 
> If, in your experience, those prompts suffice and exposing an option to customize them doesn't help with the overall performance, we can discard the possibility.

Thanks! While I haven't tested different augmentations, I am all for customizability. I have changed the ``with_augs`` parameter in ``__init__`` from a bool to a list of string augmentations. The list defaults to ``AUGS_CONST`` but can be overridden.

**Review by sayakpaul — APPROVED** [review]
Thanks so much. I will wait for the CI and then let's ship this! 🚀

**Pull request merged**
